File: dtale_desktop/frontend/src/store/actions.ts
import { Dispatch } from "react";
import { Node, Source, RootState, Settings } from "./state";

export const addSources = (sources: Source[]) => ({
  type: "ADD_SOURCES" as const,
  sources: sources,
});

export const setSelectedSource = (source: Source | null) => ({
  type: "SET_SELECTED_SOURCE" as const,
  source: source,
});

export const updateSource = (source: Source) => ({
  type: "UPDATE_SOURCE" as const,
  source: source,
});

export const setSourceUpdating = (
  sourceId: Source["id"],
  updating: boolean
) => ({
  type: "SET_SOURCE_UPDATING" as const,
  sourceId: sourceId,
  updating: updating,
});

export const updateNode = (node: Node) => ({
  type: "UPDATE_NODE" as const,
  node: node,
});

export const setNodeUpdating = (dataId: Node["dataId"], updating: boolean) => ({
  type: "SET_NODE_UPDATING" as const,
  dataId: dataId,
  updating: updating,
});

export const setOpenModal = (modal: RootState["openModal"]) => ({
  type: "SET_OPEN_MODAL" as const,
  modal: modal,
});

export const updateSettings = (settings: Settings) => ({
  type: "UPDATE_SETTINGS" as const,
  settings: settings,
});

export type NodeAction = ReturnType<typeof updateNode | typeof setNodeUpdating>;

export type SourceAction =
  | ReturnType<
      typeof addSources | typeof updateSource | typeof setSourceUpdating
    >
  | NodeAction;

export type Action =
  | ReturnType<
      typeof setSelectedSource | typeof setOpenModal | typeof updateSettings
    >
  | SourceAction;

export type ActionDispatch = Dispatch<Action>;

File: dtale_desktop/actions.py
from typing import Optional, List

from pydantic.fields import Field
from typing_extensions import Literal

from dtale_desktop.models import DataSourceSerialized, Node
from dtale_desktop.pydantic_utils import BaseApiModel
from dtale_desktop.settings import settings
from dtale_desktop.websocket_connections import websocket_connection_manager


class _Action(BaseApiModel):
    async def broadcast(self, exclude: Optional[List[int]] = None) -> None:
        await websocket_connection_manager.broadcast(
            _Message(payload=self).json(), exclude=exclude
        )


class _Message(BaseApiModel):
    type_: str = Field(default="action", alias="type")
    payload: _Action


class UpdateSettings(_Action):
    type_: Literal["UPDATE_SETTINGS"] = Field("UPDATE_SETTINGS", alias="type")
    settings: settings.Serialized


class AddSources(_Action):
    type_: Literal["ADD_SOURCES"] = Field("ADD_SOURCES", alias="type")
    sources: List[DataSourceSerialized]


class UpdateSource(_Action):
    type_: Literal["UPDATE_SOURCE"] = Field("UPDATE_SOURCE", alias="type")
    source: DataSourceSerialized


class SetSourceUpdating(_Action):
    type_: Literal["SET_SOURCE_UPDATING"] = Field("SET_SOURCE_UPDATING", alias="type")
    source_id: str
    updating: bool = True


class UpdateNode(_Action):
    type_: Literal["UPDATE_NODE"] = Field("UPDATE_NODE", alias="type")
    node: Node


class SetNodeUpdating(_Action):
    type_: Literal["SET_NODE_UPDATING"] = Field("SET_NODE_UPDATING", alias="type")
    data_id: str
    updating: bool = True
